### PR TITLE
Documentation on supporting new services

### LINKED ISSUE #819: Write docs on steps necessary for supporting a new service.
This readme should contain items such as (but not limited to):

* Splitting code into distinct SPI and API layer
* Writing local test helpers if an emulator doesn't already exist (and discuss standards for a test helper/emulator)
* Documentation (package info, step-by-step guide, other docs that need to be updated)
* Adding example code to gcloud-java-examples
* Making sure any versioned jars/docs/etc are included in the update_pom_version.sh script
* Add the example to the list of examples in gcloud-java-examples/pom.xml in the assembler plugin
* Look at the SLA to determine the appropriate default retry settings (see #857)
* Be aware that library version in READMEs is auto-updated by utilities/update_docs_version.sh when releasing.
* Mention that code should be added to [gcloud-java-examples/test-apps](https://github.com/GoogleCloudPlatform/gcloud-java-examples/tree/master/test-apps) for each new service. 

This doc can be linked from the CONTRIBUTING docs page.

### PR DESCRIPTION
**Pull request by ajkannan**
Fixes #819

### CODE REVIEW

File: ADDING_NEW_CLIENTS.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+* A command line example application.
+* Documentation, which is comprised of READMEs, Javadoc, and code snippets.
+
+### Components of a new client library
+
+#### API layer
+
+Before starting work on the API layer, write a design document and provide sample API code, either in the design document or as a pull request tagged with the "don't merge" label.  As part of the design process, be sure to examine the Google Cloud service API and any implementations provided in other gcloud-* language libraries.  Solicit feedback from other contributors to the repository.
+
+When possible, make classes immutable, providing builders when necessary.  Commonly-used classes that contain metadata should also contain a subclass that provides functions on that metadata.  For example, see `BlobInfo` (the metadata class) and `Blob` (the functional class).  The builders for both objects should implement a common interface or abstract class, and the subclass should delegate to the metadata class builder.  Make model object classes serializable.  Also, make classes final when possible, except when the class contains functionality that cannot be fully tested by users without mocking the object.
```
Comment (mziccard) [inline ADDING_NEW_CLIENTS.md]: Some comments on this paragraph:
- We could link to `BlobInfo` and `Blob` classes.
- "the subclass should delegate to the metadata class builder" -> "the subclass builder should delegate to the metadata class builder"
- We could mention that if the class cannot be made final then `hashCode` and `equals` should be final

File: ADDING_NEW_CLIENTS.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+Notes/reminders:
+* API layer classes should be located in the package `com.google.cloud.servicename`, where "servicename" corresponds to the name of the Cloud service.
+* Override the `ServiceOptions.defaultRetryParams()` method in your service's options class to align with the Service Level Agreement (SLA) given by the underlying service.  See #857 and #860 for context.
+* See conventions about overriding the `equals` and `hashCode` methods in the discussion of #892.
+* While not all fields for model objects need to be exposed to the user, `gcloud-java` clients should get and set all relevant fields when making RPC calls to the Cloud service's API.  For example, since the `parent` field of Cloud Resource Manager Project objects is in alpha (at the time this is written) and not available to most users, `gcloud-java-resourcemanager` gets and sets the parent when interacting with the Cloud Resource Manager, but does not expose the parent to users. This avoids the user inadvertently attempting to unset the parent when updating a project.
```
Comment (mziccard) [inline ADDING_NEW_CLIENTS.md]: I feel like in "This avoids the user inadvertently attempting" a conjunction is needed.

File: ADDING_NEW_CLIENTS.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+Notes/reminders:
+* Clearly document which APIs must be enabled in the Developers Console's API Manager.
+* Versioning in documentation is automatically updated by the script `utilities/update_docs_version.sh`.  Be sure to examine that script to make sure any version-dependent documentation will be updated properly upon release.
+
+### Workflow
+
+New services should be created in a branch based on `master`.  The branch name should include "alpha".  For example, while developing `gcloud-java-pubsub`, all Pub/Sub related work should be done in `pubsub-alpha`.  All code should be submitted through pull requests from a branch on a forked repository.  Limiting pull request size is very helpful for reviewers.  All code that is merged into the branch should be standalone and well-tested.  Any todo comments in the code should have an associated Github issue number for tracking purposes.  You should periodically pull updates from the master branch, especially if there are project-wide updates or if relevant changes have been made to the core utilities library, `gcloud-java-core`.
```
Comment (mziccard) [inline ADDING_NEW_CLIENTS.md]: should include "alpha" -> should include the "-alpha" suffix?

File: ADDING_NEW_CLIENTS.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+Create at least two milestones (stable and future) for your service and an issue tag with the service name.  Create issues for any to-do items and tag them appropriately.  This keeps an up-to-date short-term to-do list and also allows for longer term roadmaps.
+
+Be sure you've configured the base folder's `pom.xml` correctly.
+  * Add your module to the base directory's `pom.xml` file under the list of modules.
+  * Add your module to the javadoc packaging settings.  See PR #802 for an example.
+  * Add your example to the assembler plugin.  PR #839 includes examples of using appassembler.
```
Comment (mziccard) [inline ADDING_NEW_CLIENTS.md]: Appassembler is not used yet right? Should we comment this?

File: ADDING_NEW_CLIENTS.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+* A test helper class, which allows users to easily interact with a local emulator (if possible).  If there is no emulator available and the service is too complex to create a mock, then this helper should facilitate separation of test data from other user data and enable easy cleanup.
+* Tests, including unit tests and integration tests.
+* A command line example application.
+* Documentation, which is comprised of READMEs, Javadoc, and code snippets.
+
+### Components of a new client library
+
+#### API layer
+
+Before starting work on the API layer, write a design document and provide sample API code, either in the design document or as a pull request tagged with the "don't merge" label.  As part of the design process, be sure to examine the Google Cloud service API and any implementations provided in other gcloud-* language libraries.  Solicit feedback from other contributors to the repository.  Client libraries should be low-level while minimizing boilerplate code needed by API users.  The library should also be flexible enough to be used by higher-level libraries.  For example, Objectify should be able to use `gcloud-java-datastore`.
```
Comment (mderka) [inline ADDING_NEW_CLIENTS.md]: We should mention that the design document is essentially showing the intended interaction of the user with the new gcloud-java module. It may not be obvious for a person who never saw an example.

File: ADDING_NEW_CLIENTS.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+The library should contain:
+* A subclass of the [`ServiceOptions`](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-core/src/main/java/com/google/cloud/ServiceOptions.java) class. The `ServiceOptions` class contains important information for communicating with the Google Cloud service, such as the project ID, authentication, and retry handling.  Subclasses should add/override relevant methods as necessary.  Example: [`DatastoreOptions`](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-datastore/src/main/java/com/google/cloud/datastore/DatastoreOptions.java)
```
Comment (mderka) [inline ADDING_NEW_CLIENTS.md]: Period.

File: ADDING_NEW_CLIENTS.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+* An interface extending the [`Service`](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-core/src/main/java/com/google/cloud/Service.java) interface (example: [`Datastore`](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-datastore/src/main/java/com/google/cloud/datastore/Datastore.java)) and an implentation of that interface (example: [`DatastoreImpl`](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java)).
```
Comment (mderka) [inline ADDING_NEW_CLIENTS.md]: Or get rid of this period.

File: ADDING_NEW_CLIENTS.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+* Exception handling: we have the notion of retryable vs non-retryable exceptions and idempotent vs non-idempotent exceptions.  These are encapsulated by `BaseServiceException`.  Retryable error codes should be listed in the client's subclass of `BaseServiceException`, specifying whether exception is idempotent.  The subclass should also provide methods to translate the exceptions given by the underlying autogeneraged client library.  The [`ExceptionHandler`](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-core/src/main/java/com/google/cloud/ExceptionHandler.java) class intercepts and decides how to handle exceptions (based on the set of retryable error codes) when making RPC calls.
```
Comment (mderka) [inline ADDING_NEW_CLIENTS.md]: We should mention that a retryable exception is an exception where the operation can (and will be!) retried without any possible detrimental effect, and where it makes sense (e.g., invalid data are not retryable). Also, we should explain idempotent. Perhaps also mention the retry helper.

File: ADDING_NEW_CLIENTS.md [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+* Batching: The [`BatchResult`](https://github.com/GoogleCloudPlatform/gcloud-java/blob/dns-alpha-batch/gcloud-java-core/src/main/java/com/google/cloud/BatchResult.java) class provides a simple way for users to combine RPC calls for performance enhancement.  Clients for services that provide batching should provide a batch class that contains methods similar to the methods in the `Service.java` subclass, but instead return a subclass of `BatchResult`.  Also provide an SPI-layer class to collect batch requests and submit the batch.
+
+Notes/reminders:
+* API layer classes should be located in the package `com.google.cloud.servicename`, where "servicename" corresponds to the name of the Cloud service.
+* Override the `ServiceOptions.defaultRetryParams()` method in your service's options class to align with the Service Level Agreement (SLA) given by the underlying service.  See #857 and #860 for context.
```
Comment (mderka) [inline ADDING_NEW_CLIENTS.md]: The issues do not convert to links here.

File: ADDING_NEW_CLIENTS.md [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+* See conventions about overriding the `equals` and `hashCode` methods in the discussion of #892.
+* While not all fields for model objects need to be exposed to the user, `gcloud-java` clients should get and set all relevant fields when making RPC calls to the Cloud service's API.  For example, since the `parent` field of Cloud Resource Manager Project objects is in alpha (at the time this is written) and not available to most users, `gcloud-java-resourcemanager` gets and sets the parent when interacting with the Cloud Resource Manager, but does not expose the parent to users.  As a result, the user won't accidentally unset the parent when updating a project.
+* Be aware of differences in "update" behavior and name update/replace methods accordingly in your API.  See #321 for context.
```
Comment (mderka) [inline ADDING_NEW_CLIENTS.md]: The issue number does not convert to a link.

File: ADDING_NEW_CLIENTS.md [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+* Do not expose third party libraries in the API.  This has been a design choice from the beginning of the project, and all existing clients adhere to this convention.
+
```
Comment (mderka) [inline ADDING_NEW_CLIENTS.md]: Add that getters and setters should not follow the bean pattern. Automatically generated ID should be named as `generatedId()`.

File: ADDING_NEW_CLIENTS.md [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+There are three types of test helpers:
+* When a local emulator is already available, your test helper should launch that emulator and return service options to connect to that local emulator.  This enables both users and our own library to run unit tests easily.  An example of this type of helper is `LocalDatastoreHelper`.  Google Cloud Datastore provides a script that launches a local datastore, so `LocalDatastoreHelper` launches that script in a separate process when the user calls `start()`.
+
+* When there is no local emulator but the service is simple enough to write an emulator, you should do so.  The emulator should listen to a port for requests, process those requests, and send responses back, being as true to the actual service as possible.  Be sure to document differences between your emulator and the actual service.  Examples of this type of test helper are `LocalResourceManagerHelper` and `LocalDnsHelper`.
```
Comment (mderka) [inline ADDING_NEW_CLIENTS.md]: Perhaps recommend using lightweight dependencies?

File: ADDING_NEW_CLIENTS.md [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+  * Get service options settings.
+  * Create a test bucket with a sufficiently obscure name (to separate the bucket from any of the users other data).
+  * Clear up data left over from tests in that test bucket.
+
+
+#### Tests
+
+API-level functionality should be well-covered by unit tests.  Coders and reviewers should examine test coverage to ensure that important code paths are not being left untested.  As of now, `gcloud-java` relies on integration tests to test the SPI layer.  Unit tests for the API layer should be located in the package `com.google.cloud.servicename`.  Integration tests should be placed in a separate package, `com.google.cloud.servicename.it`, which enables us to catch method access bugs.  Unit tests for the test helper should be placed in the package `com.google.cloud.servicename.testing`.  All unit tests run for pull requests, but integration tests are only run upon merging the pull request.  We only run integration tests upon merging pull requests to avoid decrypting and exposing credentials to anybody who can create a pull request from a fork.
```
Comment (mderka) [inline ADDING_NEW_CLIENTS.md]: Mention that integration test should be prefixed with IT which excludes it from travis. Also mention that integration tests currently do not run in build except for when PRs are merged. Also, assumptions for running integration tests should be well documented, and users should take into account that the tests should run with any third party's project and not damage the data inside it.

File: ADDING_NEW_CLIENTS.md [hunk 14]
```diff
@@ -73,21 +88,20 @@ Notes/reminders:
...
+New services should be created in a branch based on `master`.  The branch name should include the suffix "-alpha".  For example, while developing `gcloud-java-pubsub`, all Pub/Sub related work should be done in `pubsub-alpha`.  All code should be submitted through pull requests from a branch on a forked repository.  Limiting pull request size is very helpful for reviewers.  All code that is merged into the branch should be standalone and well-tested.  Any todo comments in the code should have an associated Github issue number for tracking purposes.  You should periodically pull updates from the master branch, especially if there are project-wide updates or if relevant changes have been made to the core utilities library, `gcloud-java-core`.
```
Comment (mderka) [inline ADDING_NEW_CLIENTS.md]: Give a recommendation to keep PRs under a specific number of locs. Say 1500.

File: ADDING_NEW_CLIENTS.md [hunk 15]
```diff
@@ -73,21 +88,20 @@ Notes/reminders:
...
 
 Create at least two milestones (stable and future) for your service and an issue tag with the service name.  Create issues for any to-do items and tag them appropriately.  This keeps an up-to-date short-term to-do list and also allows for longer term roadmaps.
 
 Be sure you've configured the base folder's `pom.xml` correctly.
```
Comment (mderka) [inline ADDING_NEW_CLIENTS.md]: Keep dependencies in alphabetical order.

**Comment by mderka** [comment]
Very nice. I am done.

File: ADDING_NEW_CLIENTS.md (hunk 5 above)
Comment (ajkannan) [inline ADDING_NEW_CLIENTS.md]: I think this is covered in "provide sample API code", but if that's not clear, let me know and I'll add a sentence on that.

File: ADDING_NEW_CLIENTS.md (hunk 9 above)
Comment (ajkannan) [inline ADDING_NEW_CLIENTS.md]: Yeah, I think that's because this README is in my fork rather than merged into master.  Once it's merged, it should work.

File: ADDING_NEW_CLIENTS.md (hunk 14 above)
Comment (ajkannan) [inline ADDING_NEW_CLIENTS.md]: I'll probably leave this out because it's hard to come up with a specific recommendation (since I won't be reviewing pull requests too much anymore) and sometimes 1500 lines of pure code is way too much whereas 1500 lines of boilerplate code (including licences, etc) is totally fine.

**Comment by ajkannan** [comment]
Thanks @mderka!  I pushed a commit with the changes.

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -43,9 +43,9 @@ The feature must work fully on Java 7 and above.
 The feature must not add unnecessary dependencies (where "unnecessary" is of course subjective,
 but new dependencies should be discussed).
 
-Adding a New Client Library
----------------------------
-See [ADDING_NEW_CLIENTS](./ADDING_NEW_CLIENTS.md) for guidelines on how to add a new client library to `gcloud-java`.
+Adding Support for a New Service
+--------------------------------
+See [SUPPORTING_NEW_SERVICES](./SUPPORT_NEW_SERVICES.md) for guidelines on how to add a new client library to `gcloud-java`.
```
Comment (ajkannan) [inline CONTRIBUTING.md]: Note: fix link to be SUPPORTING rather than SUPPORT

File: SUPPORTING_NEW_SERVICES.md [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+
+Create at least two milestones (stable and future) for your service and an issue tag with the service name.  Create issues for any to-do items and tag them appropriately.  This keeps an up-to-date short-term to-do list and also allows for longer term roadmaps.
+
+Be sure you've configured the base folder's `pom.xml` correctly.
+  * Add your module to the base directory's `pom.xml` file under the list of modules (in alphabetical order).
+  * Add your module to the javadoc packaging settings.  See PR #802 for an example.
+  <!--- (TODO: uncomment when gcs-nio branch is merged into master) * Add your example to the assembler plugin.  PR #839 includes examples of using appassembler. --->
```
Comment (ajkannan) [inline SUPPORTING_NEW_SERVICES.md]: Note: comment this out for reals (appears in "View")

File: SUPPORTING_NEW_SERVICES.md [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+
+Create at least two milestones (stable and future) for your service and an issue tag with the service name.  Create issues for any to-do items and tag them appropriately.  This keeps an up-to-date short-term to-do list and also allows for longer term roadmaps.
+
+Be sure you've configured the base folder's `pom.xml` correctly.
+  * Add your module to the base directory's `pom.xml` file under the list of modules (in alphabetical order).
+  * Add your module to the javadoc packaging settings.  See PR #802 for an example.
+  [comment]: # (TODO: uncomment when gcs-nio branch is merged into master) * Add your example to the assembler plugin.  PR #839 includes examples of using appassembler.
```
Comment (ajkannan) [inline SUPPORTING_NEW_SERVICES.md]: To fix: figure out how to actually comment this line out.

File: SUPPORTING_NEW_SERVICES.md [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+
+* Batching: The [`BatchResult`](https://github.com/GoogleCloudPlatform/gcloud-java/blob/dns-alpha-batch/gcloud-java-core/src/main/java/com/google/cloud/BatchResult.java) class provides a simple way for users to combine RPC calls for performance enhancement.  API for services that support batching should implement a batch class that contains methods similar to the methods in the `Service.java` subclass, but instead return a subclass of `BatchResult`.  Also provide an SPI-layer class to collect batch requests and submit the batch.
```
Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: I find the "but instead .." part confusion. Instead of what? I think we can say "but instead of the original return type..." or just say "..`Service.java` subclass. Batch operations' return type should be be a subclass of `BatchResult`".

Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: We could elaborate on batchig a little more. We can say that the a Batch instance should be created by the service api (preferably using a `batch()` method) and that the batch submission is done via the batch instance itself (preferable using a `submit()` method).

File: SUPPORTING_NEW_SERVICES.md [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+
+Notes/reminders:
+* API layer classes should be located in the package `com.google.cloud.servicename`, where "servicename" corresponds to the name of the Cloud service.
```
Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: should Cloud be capitalized  here? (should it be Google Cloud). Fine as is if you think what you have is correct.

File: SUPPORTING_NEW_SERVICES.md [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+* Override the `ServiceOptions.defaultRetryParams()` method in your service's options class to align with the Service Level Agreement (SLA) given by the underlying service.  See #857 and #860 for context.
```
Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: Regarding overriding ServiceOptions method, should we mention overriding `projectIdRequired` to return false when service does not require a project to be associated with specific project?

File: SUPPORTING_NEW_SERVICES.md [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+* A runtime exception class that extends [`BaseServiceException`](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-core/src/main/java/com/google/cloud/BaseServiceException.java), which is used to wrap service-related exceptions.  Example: [`DatastoreException`](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-datastore/src/main/java/com/google/cloud/datastore/DatastoreException.java)
+* Classes representing model objects and request-related options.  Model objects that correspond to service resources should have a subclass that provides functions related to that resource.  For example, see [`BlobInfo`](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-storage/src/main/java/com/google/cloud/storage/BlobInfo.java) (the metadata class) and [`Blob`](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-storage/src/main/java/com/google/cloud/storage/Blob.java) (the functional class).  The builders for both objects should implement a common interface or abstract class, and the functional subclass builder should delegate to the metadata class builder.
+
```
Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: Should we describe that for optional params of a service operation we chose to use this pattern:
- Pass an XXXOption class as varargs.
- Option class provide static method to create an instance with specific option settings.

A common option is to request that only specific fields should be included in the response.
Typically such option is created via a `fields(...)` method which accepts a vargarg of XXXField
enum (that implements [FieldSelector](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-core/src/main/java/com/google/cloud/FieldSelector.java)).

File: SUPPORTING_NEW_SERVICES.md [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+In general, make classes immutable whenever possible, providing builders as necessary.  Make model object classes `java.util.Serializable`.  Prefer making classes final, with the following exceptions: (1) functional objects and (2) classes in which the user cannot set all attributes.  If a class cannot be made final, then `hashCode` or `equals` overrides should be made final if possible.
+
+`gcloud-java-core` provides functionality for code patterns used across `gcloud-java` libraries.  The following are some important core concepts:
```
Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: core also provides IAM classes that should be used if the Service supports IAM (and maybe we should discuss more about how to use it).

File: SUPPORTING_NEW_SERVICES.md [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+* See conventions about overriding the `equals` and `hashCode` methods in the discussion of #892.
+* While not all fields for model objects need to be exposed to the user, `gcloud-java` services should get and set all relevant fields when making RPC calls to the Cloud service's API.  For example, since the `parent` field of Cloud Resource Manager Project objects is in alpha (at the time this is written) and not available to most users, `gcloud-java-resourcemanager` gets and sets the parent when interacting with the Cloud Resource Manager, but does not expose the parent to users.  As a result, the user won't accidentally unset the parent when updating a project.
+* Be aware of differences in "update" behavior and name update/replace methods accordingly in your API.  See #321 for context.
+* Do not expose third party libraries in the API.  This has been a design choice from the beginning of the project, and all existing `gcloud-java` services adhere to this convention.
+* Member variable getters and builder setters should not use the prefixes "get" or "set".
```
Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: I would be more specific and say something like "the JavaBean get/set prefix style"

File: SUPPORTING_NEW_SERVICES.md [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+
+* When there is no local emulator but the service is simple enough to write an emulator, you should do so.  The emulator should listen to a port for requests, process those requests, and send responses back, being as true to the actual service as possible.  Dependencies in this mock should be as lightweight as possible.  Be sure to document differences between your emulator and the actual service.  Examples of this type of test helper are `LocalResourceManagerHelper` and `LocalDnsHelper`.
+
+* When there is no local emulator and the service is too complex to write a solid emulator, the test helper should contain methods to get options and separate test data from other user data.  `RemoteStorageHelper` is an example of this type of test helper, since there is no local emulator for Google Cloud Storage (at the time that this is written) and because the Google Cloud Storage API is complex.  `RemoteStorageHelper` has methods to:
```
Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: > and separate test data from other user data.

and to help with isolating the test data from the production data.

Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: > at the time that this is written

Strange, for some reason "at the time this **was** written" sounds better to me however Google proves me wrong.

"at the time this is written" -> About 4,080,000 results
"at the time this was written" -> About 849,000 results

File: SUPPORTING_NEW_SERVICES.md [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+  * Get service options settings.
+  * Create a test bucket with a sufficiently obscure name (to separate the bucket from any of the users other data).
+  * Clean up data left over from tests in that test bucket.
+
+#### Tests
+
+API-level functionality should be well-covered by unit tests.  Coders and reviewers should examine test coverage to ensure that important code paths are not being left untested.  As of now, `gcloud-java` relies on integration tests to test the SPI layer.  Unit tests for the API layer should be located in the package `com.google.cloud.servicename`.  Integration tests should be placed in a separate package, `com.google.cloud.servicename.it`, which enables us to catch method access bugs.  Unit tests for the test helper should be placed in the package `com.google.cloud.servicename.testing`.  All unit tests run for pull requests, but integration tests are only run upon merging the pull request.  We only run integration tests upon merging pull requests to avoid decrypting and exposing credentials to anybody who can create a pull request from a fork.  To separate integration tests from unit tests,prefix integration test file names with "IT" (e.g. `ITDnsTest`).
```
Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: | tests,prefix 

is that comma necessary? If so, should it have a space after?

File: SUPPORTING_NEW_SERVICES.md [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+
+Simple service-related tests should be added to [GoogleCloudPlatform/gcloud-java-examples](https://github.com/GoogleCloudPlatform/gcloud-java-examples/tree/master/test-apps).  To test releases and platform-specific bugs, it's valuable to deploy the apps in that repository on App Engine, Compute Engine, and from your own desktop.
+
+#### Example application
+
+The example application should be a simple command line interface for the service.  It should use common library use patterns so that users see good examples of how to use `gcloud-java` when viewing the source code.  Be sure to keep the examples up to date if/when there are updates that make the API cleaner and more concise.  See examples of applications under the `gcloud-java-examples` folder.  The example application should be in the package `com.google.cloud.examples.servicename`.
```
Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: > It should use common library use patterns

usage pattern?

Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: > users see good examples

users can have a good example of how to use the service... ?

File: SUPPORTING_NEW_SERVICES.md [hunk 28]
```diff
@@ -0,0 +1,108 @@
...
+* Do not expose third party libraries in the API.  This has been a design choice from the beginning of the project, and all existing `gcloud-java` services adhere to this convention.
+* Member variable getters and builder setters should not use the prefixes "get" or "set".
+* Any service-generated IDs for model objects should be named `generatedId()`.
+
+#### SPI layer
+
+The SPI layer classes should be located in the package `com.google.cloud.servicename.spi`.  In most cases, the SPI layer should contain at least three classes:
+* An RPC factory interface (allows for the implementation to be loaded via the `java.util.ServiceLoader`).
+* An RPC interface that contains all RPC methods.
+* A default RPC implementation.
```
Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: Should we have a reminder to include the new spi in [here](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/pom.xml#L389)?

File: SUPPORTING_NEW_SERVICES.md [hunk 29]
```diff
@@ -0,0 +1,108 @@
...
+
+#### Example application
+
+The example application should be a simple command line interface for the service.  It should use common library use patterns so that users see good examples of how to use `gcloud-java` when viewing the source code.  Be sure to keep the examples up to date if/when there are updates that make the API cleaner and more concise.  See examples of applications under the `gcloud-java-examples` folder.  The example application should be in the package `com.google.cloud.examples.servicename`.
+
+#### Documentation
+
+* Include a summary of the service and code snippets on the main repository's README.  These snippets should be simple and cover a few common usage patterns.  The README snippets should also be added to `gcloud-java-examples` in the package `com.google.cloud.examples.servicename.snippets`.  Placing snippet code in the repository ensures that the snippet code builds when Travis CI is run.  For this purpose, README snippets and the snippet code in `gcloud-java-examples` should be kept in sync.  As of yet, we do not have unit tests for snippets, so the snippets should be tested periodically, especially after any relevant library updates.
```
Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: > should be kept in sync

Should we mention #753?

File: SUPPORTING_NEW_SERVICES.md [hunk 30]
```diff
@@ -0,0 +1,108 @@
...
+
+Notes/reminders:
+* Clearly document which APIs must be enabled in the Developers Console's API Manager.
+* Versioning in documentation is automatically updated by the script `utilities/update_docs_version.sh`.  Be sure to examine that script to make sure any version-dependent documentation will be updated properly upon release.
+
+### Workflow
+
+New services should be created in a branch based on `master`.  The branch name should include the suffix "-alpha".  For example, while developing `gcloud-java-pubsub`, all Pub/Sub related work should be done in `pubsub-alpha`.  All code should be submitted through pull requests from a branch on a forked repository.  Limiting pull request size is very helpful for reviewers.  All code that is merged into the branch should be standalone and well-tested.  Any todo comments in the code should have an associated Github issue number for tracking purposes.  You should periodically pull updates from the master branch, especially if there are project-wide updates or if relevant changes have been made to the core utilities library, `gcloud-java-core`.
```
Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: > You should periodically pull updates from the master branch
 
Should we mention that such updated to the branch should be done in its own PR?

File: SUPPORTING_NEW_SERVICES.md [hunk 31]
```diff
@@ -0,0 +1,108 @@
...
+
+Create at least two milestones (stable and future) for your service and an issue tag with the service name.  Create issues for any to-do items and tag them appropriately.  This keeps an up-to-date short-term to-do list and also allows for longer term roadmaps.
+
+Be sure you've configured the base folder's `pom.xml` correctly.
+  * Add your module to the base directory's `pom.xml` file under the list of modules (in alphabetical order).
+  * Add your module to the javadoc packaging settings.  See PR #802 for an example.
```
Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: Oh, ignore my comment above regarding spi and the note to update the main pom as I see that you are mentioning it here.

File: SUPPORTING_NEW_SERVICES.md (hunk 18 above)
Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: comment does not look good (it is visible) in the preview mode. is it correct?

File: SUPPORTING_NEW_SERVICES.md [hunk 32]
```diff
@@ -0,0 +1,108 @@
...
+  [comment]: # (TODO: uncomment when gcs-nio branch is merged into master) * Add your example to the assembler plugin.  PR #839 includes examples of using appassembler.
+
+When your service is complete, contact the service owners to get a review.  The primary purpose of this review is to make sure that the gcloud-java service interacts with the Cloud service properly.  Present the reviewers with a link to the Github repository, as well as your (updated) design document that details the API.
+
+### Closing remarks
+
+* Efforts should be made to maintain the current style of the repository and a consistent style between gcloud-java services.
+  * We anticipate that people will sometimes use multiple `gcloud-java` services, so we don't want differences in conventions from library to library.  Look at existing `gcloud-java` services to see coding and naming conventions.
```
Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: >  from library to library.

from one service API to another ?

File: SUPPORTING_NEW_SERVICES.md [hunk 33]
```diff
@@ -0,0 +1,108 @@
...
+  * Codacy is configured to report on pull requests about style issues.  Whenever possible, those comments should be addressed.  Coders and reviewers should also run a linter on pull requests, because the Codacy tool may not catch all style errors.
+* When weighing which services to add, consider that a hand-crafted `gcloud-java` service API is especially useful if it can abstract away and/or make java-idiomatic significant parts of a service's autogenerated API.
```
Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: > it can abstract away and/or make java-idiomatic significant parts of a service's autogenerated API.

Instead I would say "if it can simplify the usability of the auto-generated client"

File: SUPPORTING_NEW_SERVICES.md [hunk 34]
```diff
@@ -0,0 +1,108 @@
...
+  [comment]: # (TODO: uncomment when gcs-nio branch is merged into master) * Add your example to the assembler plugin.  PR #839 includes examples of using appassembler.
+
+When your service is complete, contact the service owners to get a review.  The primary purpose of this review is to make sure that the gcloud-java service interacts with the Cloud service properly.  Present the reviewers with a link to the Github repository, as well as your (updated) design document that details the API.
+
+### Closing remarks
+
+* Efforts should be made to maintain the current style of the repository and a consistent style between gcloud-java services.
+  * We anticipate that people will sometimes use multiple `gcloud-java` services, so we don't want differences in conventions from library to library.  Look at existing `gcloud-java` services to see coding and naming conventions.
+  * Codacy is configured to report on pull requests about style issues.  Whenever possible, those comments should be addressed.  Coders and reviewers should also run a linter on pull requests, because the Codacy tool may not catch all style errors.
```
Comment (aozarov) [inline SUPPORTING_NEW_SERVICES.md]: should we mention this [Checkstyle config](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/checkstyle.xml)?

**Comment by aozarov** [comment]
I am done with my review. Looks really good, thanks for doing that!

**Comment by ajkannan** [comment]
All comments incorporated, ready for another pass.

**Comment by aozarov** [comment]
Looks great. Feel free to squash and merge.
